### PR TITLE
DM-30027: Add DC2 data processing guide

### PR DESCRIPTION
**Pull request by mrawls**
This guide is intended as a starting place for lsst-devl and/or DP0 users to learn how to process DC2 data through an AP Pipeline. It includes command-line "pipetask run" examples as well as references to the BPS documentation to scale up a single-patch run to an entire tract.

If there is a better place for this to land, please let me know.

I'll attach a PDF of what the page currently looks like when built to the ticket.

### CODE REVIEW

File: getting-started/dc2-guide.rst [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+The "tree chain" view is a handy way to see the "chained" data products that are included with each collection, such as calibs, skymaps, refcats, etc.
+
+To see some available datasets for processing, try, e.g.,
+
+.. prompt:: bash
+
+   butler query-data-ids /repo/dc2 tract patch visit --collections='2.2i/defaults/test-med-1' --where "skymap='DC2' AND band='g' AND tract=3828 AND patch=47 AND instrument='LSSTCam-imSim'"
+
+This command should print a list of data IDs that meet the search criteria, along with their tract, patch, and visit number.
+
```
Comment (isullivan) [inline getting-started/dc2-guide.rst:44]: Could you include a comment stating which of the `--where` arguments are required?

File: getting-started/dc2-guide.rst [hunk 2]
```diff
@@ -0,0 +1,182 @@
+########################################
+Processing DC2 data with the Gen3 Butler
+########################################
+
+The data explored in this guide is simulated data, the same kind used for Rubin's Data Preview 0 (DP0).
+
+.. note::
+
+   This guide assumes the user has access to a shared "Gen3" Butler repository containing data from the Dark Energy Science Collaboration (DESC)'s Data Challenge 2 (DC2), most likely on the ``lsstdevl`` machines at NCSA.
+   This guide further assumes the user has a recently-built version of :py:mod:`lsst.distrib` from the LSST Science Pipelines.
```
Comment (isullivan) [inline getting-started/dc2-guide.rst:10]: Could you include the tag of the build you were using when you wrote this guide? E.g. `lsst-scipipe-0.6.0` with `w_2021_30`

File: getting-started/dc2-guide.rst [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+       - characterizeImage
+       - calibrate
+       - consolidateVisitSummary
+     description: >
+       Tasks to run for single frame processing that are necessary to use the good seeing selector to build coadds for use as difference imaging templates.
+
+This example pipeline imports a pipeline from :py:mod:`lsst.ap.pipe` you may `view on GitHub <https://github.com/lsst/ap_pipe/blob/master/pipelines/ApTemplate.yaml>`__.
+There are some special configurations concerning reference catalogs that must be set for this camera and/or dataset, so the example pipeline above lists the ``calibrate`` task explicitly to add custom configurations.
+
+To run this example pipeline, save it as ``ApTemplate-DC2.yaml``, choose an appropriate output collection name, and run
```
Comment (isullivan) [inline getting-started/dc2-guide.rst:89]: Suggested change:
...choose an appropriate output collection name (u/USERNAME/OUTPUT-COLLECTION-1 in the example below)...

File: getting-started/dc2-guide.rst [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+   - location: $AP_PIPE_DIR/pipelines/ApPipe.yaml
+     exclude:  # These tasks come from the ApTemplate-DC2 pipeline instead
+       - isr
+       - characterizeImage
+       - calibrate
+
+This difference imaging pipeline uses the good seeing templates we built and treats all the DP0 defaults as input "science" images.
+
+**This is a two-step process.**
+First, create an empty sqlite APDB:
```
Comment (isullivan) [inline getting-started/dc2-guide.rst:131]: Please add a comment that the APDB must be empty, so they must make a new one if they need to re-run the pipeline for any reason.

File: getting-started/dc2-guide.rst [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+When this pipeline completes, you should have difference images and an APDB with populated tables (``DiaSource``, ``DiaObject``, etc.) for multiple bands in patch 47 of tract 3829 of this dataset.
+
+Processing Data with BPS
+========================
+
+The example data processing steps above assume a relatively small data volume (a single patch), so running from the command line and using an sqlite APDB is appropriate.
+However, if you want to process larger data volumes, you'll need to use the Batch Processing System (BPS, :py:mod:`lsst.ctrl.bps`) and a PostgreSQL APDB.
+
+Describing how to set up a PostgreSQL APDB is beyond the scope of this guide. One key difference between using an sqlite APDB versus a PostgreSQL APDB is that the former is a file on disk created from scratch when running ``make_apdb.py``.
```
Comment (isullivan) [inline getting-started/dc2-guide.rst:151]: Could you provide a link to your PostgreSQL APDB notes here? Perhaps with a caveat that they are not meant to be a complete user guide.

File: getting-started/dc2-guide.rst [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+The latter requires a database to already exist, and ``make_apdb.py`` turns the PostgreSQL database's default schema into an empty APDB.
+As before, you will still need to run, e.g.,
+
+.. prompt:: bash
+
+   make_apdb.py -c db_url="postgresql://USER@DB_ADDRESS/DB_NAME"
```
Comment (isullivan) [inline getting-started/dc2-guide.rst:157]: I know it seems obvious, but you should probably state that `USER`, `DB_ADDRESS`, and `DB_NAME` all need to be replaced with the correct values.

**Review by isullivan — APPROVED** [review]
This looks great! I tried to read through it from the perspective of someone completely new to this processing, so I requested additional clarification at a few points where I could see someone new getting stuck.

**Comment by jonathansick** [comment]
@mrawls This looks good-to-go!

You can't link to this tutorial in the nav much like we did for the Getting Started tutorials.

First, in `index.rst`, add a bullet item for the tutorial here:

Screen Shot 2021-09-02 at 1 52 27 PM

The second place to make the link is more for the internal Sphinx structure, but is still important. Create a section in [`getting-started/index.rst`](https://pipelines.lsst.io/getting-started/index.html) and add the page with another toctree. Here's one way to do it:

```rst
.. _getting-started:

###############################################
Getting started with the LSST Science Pipelines
###############################################

Before you start working with the LSST Science Pipelines, these pages will help you get up and running.

.. _getting-started-tutorial:

Getting started tutorials
=========================

.. include:: /gen2tutorialdeprecation.txt

In this tutorial series we'll process a set of Hyper Suprime-Cam images into deep coadditions and source catalogs.
These tutorials will give you a feeling for data processing and analysis with the Science Pipelines.

.. toctree::
   :maxdepth: 1

   Part 1: setting up the Butler data repository. <data-setup>
   Part 2: calibrating single frames. <processccd>
   Part 3: displaying exposures and source catalogs. <display>
   Part 4: coadding images. <coaddition>
   Part 5: measuring sources. <photometry>
   Part 6: analyzing measurement catalogs in multiple bands. <multiband-analysis>

Additional tutorials
====================

.. toctree::
   :maxdepth: 1

   Processing DESC DC2 <dc2-guide>
```

**Comment by mrawls** [comment]
Thank you @jonathansick, this ought to be straightforward. I want to make sure it passes Jenkins again after I make these changes, and I'm about to head out on vacation, so I will merge this when I return. Feel free to point people at the guide if it would be useful in the meantime.

File: getting-started/dc2-guide.rst [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+         config.photoRefObjLoader.filterMap = {band: 'lsst_%s_smeared' % (band) for band in 'ugrizy'};
+   subsets:
+     singleFrameAp:
+     subset:
+       - isr
+       - characterizeImage
+       - calibrate
+       - consolidateVisitSummary
+     description: >
+       Tasks to run for single frame processing that are necessary to use the good seeing selector to build coadds for use as difference imaging templates.
```
Comment (spenczar) [inline getting-started/dc2-guide.rst:87]: This is missing one level of indentation, I think; it all ought to be folded under `singleFrameAp` if I'm not mistaken.

File: getting-started/dc2-guide.rst [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+     calibrate:
+       class: lsst.pipe.tasks.calibrate.CalibrateTask
+       config:
+         connections.astromRefCat: 'cal_ref_cat_2_2'
+         connections.photoRefCat: 'cal_ref_cat_2_2'
+         astromRefObjLoader.ref_dataset_name: 'cal_ref_cat_2_2'
+         photoRefObjLoader.ref_dataset_name: 'cal_ref_cat_2_2'
+         python: |
+         config.astromRefObjLoader.filterMap = {band: 'lsst_%s_smeared' % (band) for band in 'ugrizy'};
+         config.photoRefObjLoader.filterMap = {band: 'lsst_%s_smeared' % (band) for band in 'ugrizy'};
```
Comment (spenczar) [inline getting-started/dc2-guide.rst:78]: I think this is missing one level of indentation; I get a syntax error as-is. The block under the pipe needs to be indented beyond the depth of the `python` node.

File: getting-started/dc2-guide.rst [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+       Tasks to run for single frame processing that are necessary to use the good seeing selector to build coadds for use as difference imaging templates.
+
+This example pipeline imports a pipeline from :py:mod:`lsst.ap.pipe` you may `view on GitHub <https://github.com/lsst/ap_pipe/blob/master/pipelines/ApTemplate.yaml>`__.
+There are some special configurations concerning reference catalogs that must be set for this camera and/or dataset, so the example pipeline above lists the ``calibrate`` task explicitly to add custom configurations.
+
+To run this example pipeline, save it as ``ApTemplate-DC2.yaml``, choose an appropriate output collection name (``u/USERNAME/OUTPUT-COLLECTION-1`` in the example below), and run
+
+.. prompt:: bash
+
+   pipetask run -j 12 -b /repo/dc2 -d "band='g' AND skymap='DC2' AND tract=3829" -i 2.2i/defaults/test-med-1 -o u/USERNAME/OUTPUT-COLLECTION-1 -p ApTemplate-DC2.yaml#singleFrameAp --register-dataset-types
```
Comment (spenczar) [inline getting-started/dc2-guide.rst:96]: For what it's worth, this isn't working for me; I get a bunch of errors around brighter fatter kernel correction - the same error reported in these places:
 - https://lsstc.slack.com/archives/C4JQP6FRS/p1623342592112800
 - https://lsstc.slack.com/archives/C01FBUGM2CV/p1625579222311800
 - https://lsstc.slack.com/archives/C025P0D9SKB/p1624470090098600

---

I _think_ things work with just the magic `connections.newBFKernel: bfk` parameter. That is, I added this stanza under `tasks:`:
```yaml
isr:
  class: lsst.ip.isr.IsrTask
  config:
    connections.newBFKernel: bfk
    doBrighterFatter: True
```

It hasn't yet finished, but it has at least gotten further, so I think it's working.

File: getting-started/dc2-guide.rst [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+   make_apdb.py -c isolation_level=READ_UNCOMMITTED -c db_url="sqlite:////PATH-TO-DESIRED-APDB/ApPipeTest1.db"
+
+Note that the APDB must be empty, and it is highly recommended to make a new one each time the AP Pipeline is rerun for any reason.
+
+Second, run the pipeline:
+
+.. prompt:: bash
+
+   pipetask run -j 12 -b /repo/dc2 -d "skymap='DC2' AND tract=3829 AND patch=47" -i u/USERNAME/OUTPUT-COLLECTION-2,2.2i/defaults/DP0 -o u/USERNAME/OUTPUT-COLLECTION-3 -p ApPipe-DC2.yaml -c diaPipe:apdb.isolation_level=READ_UNCOMMITTED -c diaPipe:apdb.db_url="sqlite:////PATH-TO-DESIRED-APDB/ApPipeTest1.db" --register-dataset-types
```
[the remark this answers is not in the document]
Comment (mrawls) [reply getting-started/dc2-guide.rst:146]: Good catch - I meant to specify `test-med-1`, thanks! Also addressed the indentation problems you found and added the brighter-fatter config. User testing does wonders 😄

**Pull request merged**